src/app/modules/data-set/data-set-instance/data-set-instance.component.ts

```diff
@@ -0,0 +1,159 @@
+  }
+
+  constructor() { }
+
+  public ngOnInit(): void {
+  }
+
+  public ngOnChanges() {
+    this.selection.clear();
+    if (!this.isInstancesEmpty()) {
+      let annotatorId: number = +sessionStorage.getItem('annotatorID')!;
+      this.instances.forEach((instance, index) => this.instances[index] = new DataSetInstance(instance, annotatorId));
+      this.filtersChanged();
+    }
+    if (this.iframe) {
+      this.iframe.srcdoc = '';
+    }
+  }
+
+  public ngAfterViewChecked() {
+    this.iframe = document.getElementById('snippet') as HTMLIFrameElement;
+  }
+
+  public toggleInstanceSelection(selectedInstance: DataSetInstance) {
+    this.iframe.srcdoc = '';
+    this.previousAnnotation = null;
+    if (!this.selection.isSelected(selectedInstance)) {
+      this.selection.clear();
+    }
+    this.selection.toggle(selectedInstance);
+    if (this.selection.selected.length == 1) {
+      this.previousAnnotation = selectedInstance.annotationFromLoggedUser!;
+      this.iframe.srcdoc = this.createSrcdocFromGithubLink(selectedInstance.link);
+    }
+  }
+
+  private instanceHasSelectedInstanceType(instance: DataSetInstance): boolean {
+    switch (this.selectedInstanceType) {
+      case InstanceType.Class: {
+        return instance.type == InstanceType.Class;
+      }
+      case InstanceType.Method: { 
+        return instance.type == InstanceType.Method;
+      } 
+      case InstanceType.All: { 
+        return true;
+      } 
+    }
+  }
+
+  private instanceHasSelectedAnnotationStatus(instance: DataSetInstance): boolean {
+    switch (this.selectedAnnotationStatus) {
+      case AnnotationStatus.Annotated: {
+        return instance.hasAnnotationFromLoggedUser;
+      }
+      case AnnotationStatus.Not_Annotated: { 
+        return !instance.hasAnnotationFromLoggedUser;
+      } 
+      case AnnotationStatus.All: { 
+        return true;
+      } 
+    }
+  }
+
+  public filtersChanged(): void {
+    this.selection.clear();
+    this.showFilteredInstances();
+  }
+
+  public async addAnnotation(annotation: DataSetAnnotation) {
+    for (let i = 0; i < this.instances.length; i++) {
+      if (this.instances[i].id == this.selection.selected[0].id) {
+        this.instances[i].annotations.push(annotation);
+        this.instances[i] = new DataSetInstance(this.instances[i]);
+        break;
+      }
+    }
+    this.filtersChanged();
+  }
+
+  public async changeAnnotation(annotation: DataSetAnnotation) {
+    for (let i = 0; i < this.instances.length; i++) {
+      if (this.instances[i].id == this.selection.selected[0].id) {
+        for (let j = 0; j < this.instances[i].annotations.length; j++) {
+          if (this.instances[i].annotations[j].id == annotation.id) {
+            this.instances[i].annotations[j] = annotation;
+            break;
+          }
+        }
+        this.instances[i] = new DataSetInstance(this.instances[i]);
+        break;
+      }
+    }
+    this.filtersChanged();
+  }
```

Comment:
This method is somewhere diffucult to understand due to several for loops and ifs. 

`var i = this.instances.findIndex(i => i.id == this.selection.selected[0].id);`
 `var j = this.instances[i].annotations.findIndex(a => a.id == annotation.id);`
    `this.instances[i].annotations[j] = annotation;`
`    this.instances[i] = new DataSetInstance(this.instances[i]);`
 `   this.filtersChanged();`
 
With this solution, we reduce the length and complexity of the method. Just check if this code works properly (in case I made a mistake).